File: app/src/main/java/com/savvasdalkitsis/gameframe/draw/view/PaletteView.java
package com.savvasdalkitsis.gameframe.draw.view;

import android.content.Context;
import android.support.v7.widget.GridLayout;
import android.util.AttributeSet;

import com.savvasdalkitsis.gameframe.draw.model.Palette;

public class PaletteView extends GridLayout {

    private static final int SWATCH_COUNT = 20;
    private SwatchView[] swatches;
    private SwatchSelectedListener swatchSelectedListener = SwatchSelectedListener.NO_OP;

    public PaletteView(Context context) {
        super(context);
    }

    public PaletteView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public PaletteView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        int childCount = getChildCount();
        if (childCount != SWATCH_COUNT) {
            throw new IllegalStateException("Wrong number of swatches in PaletteView. Expecting " + SWATCH_COUNT + " found " + childCount);
        }
        swatches = new SwatchView[SWATCH_COUNT];
        for (int i = 0; i < SWATCH_COUNT; i++) {
            swatches[i] = (SwatchView) getChildAt(i);
            SwatchView swatch = swatches[i];
            swatch.setOnLongClickListener(view -> {
                swatchSelectedListener.onSwatchLongPressed(swatch);
                return true;
            });
        }
    }

    public void bind(Palette palette) {
        int[] colors = palette.getColors();
        for (int i = 0; i < SWATCH_COUNT; i++) {
            swatches[i].bind(colors[i]);
        }
    }

    public void deselectAllSwatches() {
        for (SwatchView swatch : swatches) {
            swatch.deselect();
        }
    }

    public void setOnSwatchSelectedListener(SwatchSelectedListener swatchSelectedListener) {
        this.swatchSelectedListener = swatchSelectedListener;
    }

    public void notifyListenerOfSwatchSelected(SwatchView swatchView) {
        swatchSelectedListener.onSwatchSelected(swatchView.getColor());
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/feature/draw/view/PaletteView.kt
package com.savvasdalkitsis.gameframe.feature.draw.view

import android.content.Context
import android.support.v7.widget.GridLayout
import android.util.AttributeSet

import com.savvasdalkitsis.gameframe.feature.draw.model.Palette

class PaletteView : GridLayout {

    private lateinit var swatches: Array<SwatchView?>
    private var swatchSelectedListener: SwatchSelectedListener? = null
    private var thumbnailMode: Boolean = false

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    override fun onFinishInflate() {
        super.onFinishInflate()
        val childCount = childCount
        if (childCount != SWATCH_COUNT) {
            throw IllegalStateException("Wrong number of swatches in PaletteView. Expecting $SWATCH_COUNT found $childCount")
        }
        swatches = arrayOfNulls(SWATCH_COUNT)
        for (i in 0 until SWATCH_COUNT) {
            swatches[i] = getChildAt(i) as SwatchView
        }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (!thumbnailMode) {
            post { this.selectFirstSwatch() }
        }
    }

    fun setThumbnailMode() {
        thumbnailMode = true
        for (swatch in swatches) {
            swatch?.let { with(it) {
                setOnClickListener(null)
                setOnLongClickListener(null)
                isClickable = false
                isLongClickable = false
            } }
        }
    }

    fun bind(palette: Palette) {
        val colors = palette.colors
        for (i in 0 until SWATCH_COUNT) {
            swatches[i]?.bind(colors[i], i)
        }
    }

    fun deselectAllSwatches() {
        for (swatch in swatches) {
            swatch?.isSelected = false
        }
    }

    fun setOnSwatchSelectedListener(swatchSelectedListener: SwatchSelectedListener) {
        this.swatchSelectedListener = swatchSelectedListener
    }

    fun notifyListenerOfSwatchSelected(swatchView: SwatchView) {
        swatchSelectedListener?.onSwatchSelected(swatchView)
    }

    fun notifyListenerOfSwatchLongClicked(swatchView: SwatchView) {
        swatchSelectedListener?.onSwatchLongPressed(swatchView)
    }

    private fun selectFirstSwatch() {
        swatches[0]?.performClick()
    }

    companion object {

        private val SWATCH_COUNT = 16
    }
}
